near-sdk/src/store/unordered_set/iter.rs

```diff
@@ -0,0 +1,412 @@
+/// See its documentation for more.
+///
+/// [`intersection`]: UnorderedSet::intersection
+pub struct Intersection<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    elements: free_list::Iter<'a, T>,
+
+    other: &'a UnorderedSet<T, H>,
+}
+
+impl<'a, T, H> Intersection<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    pub(super) fn new(set: &'a UnorderedSet<T, H>, other: &'a UnorderedSet<T, H>) -> Self {
+        Self { elements: set.elements.iter(), other }
+    }
+}
+
+impl<'a, T, H> Iterator for Intersection<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize + Clone,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        <Self as Iterator>::nth(self, 0)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, self.elements.size_hint().1)
+    }
+
+    fn count(mut self) -> usize {
+        let mut count = 0usize;
+        for element in self.elements.by_ref() {
+            if self.other.contains(element) {
+                count += 1;
+            }
+        }
+        count
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        let mut count = 0usize;
+        for element in self.elements.by_ref() {
+            if self.other.contains(element) {
+                if count == n {
+                    return Some(element);
+                }
+                count += 1;
+            }
+        }
+        None
+    }
+}
+
+impl<'a, T, H> FusedIterator for Intersection<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize + Clone,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+}
+
+/// A lazy iterator producing elements in the symmetrical difference of `UnorderedSet`s.
+///
+/// This `struct` is created by the [`symmetrical_difference`] method on [`UnorderedSet`].
+/// See its documentation for more.
+///
+/// [`symmetrical_difference`]: UnorderedSet::symmetrical_difference
+pub struct SymmetricDifference<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    iter: Chain<Difference<'a, T, H>, Difference<'a, T, H>>,
+}
+
+impl<'a, T, H> SymmetricDifference<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize + Clone,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    pub(super) fn new(set: &'a UnorderedSet<T, H>, other: &'a UnorderedSet<T, H>) -> Self {
+        Self { iter: set.difference(other).chain(other.difference(set)) }
+    }
+}
+
+impl<'a, T, H> Iterator for SymmetricDifference<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize + Clone,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+}
```

Comment:
> Wondering why you used the optimized (not sure if actually optimized or it would get compiled away) version in Intersection but not here (and Union).

The honest answer is that I just could not think of a way to represent `Intersection` as a combination of other operations. Do you know one?

> Did you confirm that there is a performance difference with Intersection? (there is same amount of storage operations, so not sure if it gets compiled away)

I did not because, as per the previous paragraph, I do not know how to implement the baseline version

> If that is faster, why is it not worth doing here as well. The inner operator doesn't get the speedups of the chained iterators nth and count operations here and would default to using next for both impls

It might indeed make sense, I do not know how good Rust compiler is with this stuff.

Reply:
> The honest answer is that I just could not think of a way to represent `Intersection` as a combination of other operations. Do you know one?

To be clear I'm just talking about the optimized versions of `nth` and `count`, not the inner type. If you don't specify `nth` here it will default to using `next` exclusively for the impls

I'm not suggesting doing a different combination, as this is even what `std` does for HashSet, I'm just suggesting either adding the `nth` and `count` operations here to be consistent if it is faster, or remove from `Intersection` if the added complexity gets compiled away (which it might). These do the same amount of storage operations, so for `Intersection` I'm not sure having `nth` and `count` actually helps.

My intuition is that `nth` and `count` custom impls aren't actually helpful and are just adding unnecessary complexity for `Intersection` and `Difference`, but that's why I'm asking if you have a reason to believe otherwise.

Reply:
Oooh, that's what you mean. Yeah, good point. I did not take a look at the `HashSet` implementation for these iterators and just intuitively thought that this way would be more efficient. Changed.